### PR TITLE
Add doc for btree_map types

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Part of #29348.

r? @steveklabnik

### CODE REVIEW

File: src/libcollections/btree/map.rs [hunk 1]
```diff
@@ -313,6 +313,10 @@ pub struct RangeMut<'a, K: 'a, V: 'a> {
 }
 
 /// A view into a single entry in a map, which may either be vacant or occupied.
+/// Can be get with [`BTreeMap`]::[`entry()`].
```
Comment (frewsxcv) [inline src/libcollections/btree/map.rs]: `s/get/got`. Though it's a bit wordy.

Maybe this? (shrug)

```
This enum is constructed from the [`entry`] method on [`BTreeMap`].
```

File: src/libcollections/btree/map.rs [hunk 2]
```diff
@@ -1867,6 +1886,17 @@ impl<'a, K: Ord, V> Entry<'a, K, V> {
...
     /// Ensures a value is in the entry by inserting the result of the default function if empty,
     /// and returns a mutable reference to the value in the entry.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::collections::BTreeMap;
+    ///
+    /// let mut map: BTreeMap<&str, usize> = BTreeMap::new();
+    /// map.entry("poneyland").or_insert_with(|| 12);
```
Comment (frewsxcv) [inline src/libcollections/btree/map.rs]: It'd be nice if this `or_insert_with` was used with a type that is move expensive to construct. Otherwise, the reader might as well use `or_insert`.

File: src/libcollections/btree/map.rs [hunk 3]
```diff
@@ -1946,43 +2027,168 @@ impl<'a, K: Ord, V> VacantEntry<'a, K, V> {
...
+    /// ```
+    /// #![feature(map_entry_recover_keys)]
+    ///
+    /// use std::collections::BTreeMap;
+    /// use std::collections::btree_map::Entry;
+    ///
+    /// let mut map: BTreeMap<&str, usize> = BTreeMap::new();
+    /// map.entry("poneyland").or_insert(12);
+    ///
+    /// match map.entry("poneyland") {
```
Comment (frewsxcv) [inline src/libcollections/btree/map.rs]: `if let`?

File: src/libcollections/btree/map.rs [hunk 4]
```diff
@@ -1946,43 +2027,168 @@ impl<'a, K: Ord, V> VacantEntry<'a, K, V> {
...
+    /// # Examples
+    ///
+    /// ```
+    /// use std::collections::BTreeMap;
+    /// use std::collections::btree_map::Entry;
+    ///
+    /// let mut map: BTreeMap<&str, usize> = BTreeMap::new();
+    /// map.entry("poneyland").or_insert(12);
+    ///
+    /// match map.entry("poneyland") {
```
Comment (frewsxcv) [inline src/libcollections/btree/map.rs]: `if let`?

File: src/libcollections/btree/map.rs [hunk 5]
```diff
@@ -313,6 +313,10 @@ pub struct RangeMut<'a, K: 'a, V: 'a> {
 }
 
 /// A view into a single entry in a map, which may either be vacant or occupied.
+/// Can be get with [`BTreeMap`]::[`entry()`].
+///
+/// [`BTreeMap`]: ./struct.BTreeMap.html
```
Comment (steveklabnik) [inline src/libcollections/btree/map.rs]: all these links shouldn't have `./` in front of them

File: src/libcollections/btree/map.rs (hunk 1 above)
Comment (steveklabnik) [inline src/libcollections/btree/map.rs]: :+1:

File: src/libcollections/btree/map.rs (hunk 2 above)
Comment (steveklabnik) [inline src/libcollections/btree/map.rs]: Agreed!

File: src/libcollections/btree/map.rs (hunk 1 above)
Comment (GuillaumeGomez) [inline src/libcollections/btree/map.rs]: :+1:

File: src/libcollections/btree/map.rs (hunk 2 above)
Comment (GuillaumeGomez) [inline src/libcollections/btree/map.rs]: I didn't have any idea of such thing at the moment I wrote the example. I still don't. :-/

Comment (frewsxcv) [inline src/libcollections/btree/map.rs]: Yeah, I was going to suggest a `Vec` or `String`

Comment (GuillaumeGomez) [inline src/libcollections/btree/map.rs]: Let's go for a `String`!

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by GuillaumeGomez** [comment]
Updated.
